### PR TITLE
Add TypeScript

### PR DESCRIPTION
**Pull request by ryanio**
This PR introduces a TypeScript build pipeline with an initial configuration to accept JS files so that parts of the project can be converted at a time.

This PR:
* Runs all code through TypeScript compilation
* Replaces browserify/babel with webpack for the browser bundle
* Updates `JSDoc` to `TypeDoc`
  * note: the diff is going to be a little crazy with this switch

### CODE REVIEW

**Comment by holgerd77** [comment]
Already looks like a great start, very cool! 😄

File: tsconfig.prod.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@ethereumjs/config-tsc",
```
Comment (tomonari-t) [inline tsconfig.prod.json:2]: We can reuse `tsconfig.json`
```diff
- "extends": "./tsconfig.json"
+ "extends": "./tsconfig.json"
```

File: tsconfig.browser.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "@ethereumjs/config-tsc",
```
Comment (tomonari-t) [inline tsconfig.browser.json:2]: We can reuse `tsconfig.json`
```diff
- "extends": "./tsconfig.json"
+ "extends": "./tsconfig.json"
```

**Comment by ryanio** [comment]
@holgerd77 would you have any insight why I am getting this error when trying to run `test/blockchain/chain.js` in the browser? I don't think I was getting it before with browserify but seeing it with karma-typescript ([ci link](https://github.com/ethereumjs/ethereumjs-client/pull/128/checks?check_run_id=763505241#step:5:43)):

```
11 06 2020 14:59:50.576:INFO [Firefox 77.0 (Mac OS 10.15)]: Connected on socket 5JHHSyl9n6nn1XVtAAAB with id 93750766
Firefox 77.0 (Mac OS 10.15) LOG: 'TAP version 13'
Firefox 77.0 (Mac OS 10.15) LOG: '# [Chain]'
Firefox 77.0 (Mac OS 10.15) LOG: '# should test object creation without logger'
Firefox 77.0 (Mac OS 10.15) LOG: 'ok 1 should be equal'
Firefox 77.0 (Mac OS 10.15) LOG: '# should test blockchain DB is initialized'
Firefox 77.0 (Mac OS 10.15) LOG: 'ok 2 read value matches written value'
Chrome Headless 83.0.4103.97 (Mac OS 10.15.4) LOG: '# should retrieve chain properties'

Chrome Headless 83.0.4103.97 (Mac OS 10.15.4) ERROR
  Uncaught AssertionError: The field transactionsTrie must have byte length of 32
  at /var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:120352:5

  AssertionError [ERR_ASSERTION]: The field transactionsTrie must have byte length of 32
      at module.exports.setter [as transactionsTrie] (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:36993:17)
      at /var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:37030:39
      at Array.forEach (<anonymous>)
      at Object.exports.defineProperties (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:37029:18)
      at new module.exports (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:8197:9)
      at /var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:122394:23
      at Blockchain._lookupByHashNumber (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:122376:3)
      at Blockchain._getHeader (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:122387:8)
      at header (/var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:121837:14)
      at /var/folders/t2/2nshsj8d65313lvzptbth0p80000gn/T/karma-typescript-bundle-64057MXfndm5lja4i.js:29871:24
```

**Comment by holgerd77** [comment]
@ryanio sorry for the late answer: maybe that's some newly added assertion recently merged in to the VM repository in the block or blockchain library?

Wonder if @jochem-brouwer added something like this somewhere, but just a blurry association :smile:, might be wrong.

**Comment by ryanio** [comment]
@holgerd77 np ok thanks I'll keep looking further, I was thinking it might also be related to some new strictness in ethereumjs-util as that error comes from validation in `defineProperties`.

**Comment by jochem-brouwer** [comment]
Hey @holgerd77 none of my code has been merged so far, so that can't be the culprit! =)

**Comment by holgerd77** [comment]
@jochem-brouwer ah, you're forgetting https://github.com/ethereumjs/ethereumjs-vm/pull/733 🙂, but there is nothing in it along to fit here. Anyhow, just an association, not sure where it was rooted. 😋  So we need to continue digging into things here.

**Comment by holgerd77** [comment]
@ryanio Ah, wasn't aware any more that this here is so "ready" ready! 😄  Is this still in the state of being somewhat mergeable (apart from the tiny README conflict from above), or has something aged or changed along the way?

**Comment by ryanio** [comment]
just resolved the readme conflict. yes it's been ready, we've just been focused on other things :) it would be good to get this merged. let me know if you see anything to address while reviewing.

File: package.json [hunk 3]
```diff
@@ -106,6 +105,13 @@
...
     "testdouble": "^3.8.2",
     "testdouble-timers": "^0.1.1",
-    "tmp": "~0.0.33"
+    "tmp": "~0.0.33",
+    "ts-loader": "^7.0.5",
+    "ts-node": "^8.10.2",
+    "typedoc": "next",
+    "typedoc-plugin-markdown": "^2.3.1",
+    "typescript": "^3.9.5",
+    "webpack": "^4.43.0",
```
Comment (jochem-brouwer) [inline package.json:114]: Where is webpack used?

**Comment by jochem-brouwer** [comment]
I am a bit confused here, seems that all files in `lib` are still `js`?

File: package.json (hunk 3 above)
Comment (ryanio) [reply package.json:114]: for `npm run bundle` which creates `dist/bundle.js`

**Comment by ryanio** [comment]
@jochem-brouwer see opening line of this PR's description:

> This PR introduces a TypeScript build pipeline with an initial configuration to accept JS files so that parts of the project can be converted at a time.

This PR at the moment does not convert any existing JS files to TS, that work is intended for subsequent PRs.

**Comment by jochem-brouwer** [comment]
Woops. Should read better 😅 

Will approve but have not checked the updates to the docs.

**Review by jochem-brouwer — APPROVED** [review]

**Pull request merged**